// @hashgraph/did-sdk-js/dist/identity/hcs/vc/credential-subject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CredentialSubject = void 0;
const hcs_vc_document_json_properties_1 = require("./hcs-vc-document-json-properties");
class CredentialSubject {
    getId() {
        return this.id;
    }
    setId(id) {
        this.id = id;
    }
    // JsonClass
    toJsonTree() {
        const rootObject = {};
        rootObject[hcs_vc_document_json_properties_1.HcsVcDocumentJsonProperties.ID] = this.id;
        return rootObject;
    }
    static fromJsonTree(root, result) {
        if (!result)
            result = new CredentialSubject();
        result.id = root[hcs_vc_document_json_properties_1.HcsVcDocumentJsonProperties.ID];
        return result;
    }
    toJSON() {
        return JSON.stringify(this.toJsonTree());
    }
    static fromJson(json) {
        let result;
        try {
            const root = JSON.parse(json);
            result = this.fromJsonTree(root);
        }
        catch (e) {
            throw new Error('Given JSON string is not a valid CredentialSubject ' + e.message);
        }
        return result;
    }
}
exports.CredentialSubject = CredentialSubject;
