// @hashgraph/did-sdk-js/dist/identity/hcs/vc/issuer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Issuer = void 0;
const hcs_vc_document_json_properties_1 = require("./hcs-vc-document-json-properties");
class Issuer {
    constructor(...args) {
        this.id = args[0];
        this.name = args[1] || null;
    }
    getId() {
        return this.id;
    }
    getName() {
        return this.name;
    }
    // JsonClass
    toJsonTree() {
        if (this.name) {
            const rootObject = {};
            rootObject[hcs_vc_document_json_properties_1.HcsVcDocumentJsonProperties.ID] = this.id;
            rootObject['name'] = this.name;
            return rootObject;
        }
        return this.id;
    }
    static fromJsonTree(root, result) {
        let id, name;
        if (typeof root == "string") {
            id = root;
        }
        else {
            id = root[hcs_vc_document_json_properties_1.HcsVcDocumentJsonProperties.ID];
            name = root["name"];
        }
        if (result) {
            result.id = id;
            result.name = name;
            return result;
        }
        else {
            return new Issuer(id, name);
        }
    }
    toJSON() {
        return JSON.stringify(this.toJsonTree());
    }
    static fromJson(json) {
        let result;
        try {
            const root = JSON.parse(json);
            result = this.fromJsonTree(root);
        }
        catch (e) {
            throw new Error('Given JSON string is not a valid Issuer ' + e.message);
        }
        return result;
    }
}
exports.Issuer = Issuer;
