// @hashgraph/did-sdk-js/dist/utils/timestamp-utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimestampUtils = void 0;
const sdk_1 = require("@hashgraph/sdk");
const moment_1 = __importDefault(require("moment"));
class TimestampUtils {
    static toJSON(item, format = this.ISO) {
        const d = item.toDate();
        return (0, moment_1.default)(d).utc().format(format);
    }
    static fromJson(json, format = this.ISO) {
        const d = moment_1.default.utc(json, format).toDate();
        return sdk_1.Timestamp.fromDate(d);
    }
    static now() {
        return sdk_1.Timestamp.fromDate(new Date());
    }
    static equals(a, b) {
        if (a == b) {
            return true;
        }
        if (!a || !b) {
            return false;
        }
        return a.seconds.equals(b.seconds) && a.nanos.equals(b.nanos);
    }
    static lessThan(a, b) {
        if (a == b) {
            return false;
        }
        if (!a || !b) {
            return false;
        }
        if (a.seconds.equals(b.seconds)) {
            return a.nanos.lessThan(b.nanos);
        }
        a.seconds.lessThan(b.seconds);
    }
}
exports.TimestampUtils = TimestampUtils;
TimestampUtils.ISO = "YYYY-MM-DDTHH:mm:ss.SSS[Z]";
TimestampUtils.ISO8601 = "YYYY-MM-DDTHH:mm:ss[Z]";
