// @hashgraph/did-sdk-js/dist/identity/hcs/address-book.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddressBook = void 0;
const sdk_1 = require("@hashgraph/sdk");
/**
 * Appnet's address book for HCS identity network.
 */
class AddressBook {
    /**
     * Converts an address book JSON string into address book object.
     *
     * @param json              Address book JSON file.
     * @param addressBookFileId FileId of this address book in Hedera File Service.
     * @return The {@link AddressBook}.
     */
    static fromJson(json, addressBookFileId) {
        const result = new AddressBook();
        const item = JSON.parse(json);
        result.appnetName = item.appnetName;
        result.didTopicId = item.didTopicId;
        result.vcTopicId = item.vcTopicId;
        result.appnetDidServers = item.appnetDidServers;
        if (typeof addressBookFileId === 'string') {
            result.setFileId(sdk_1.FileId.fromString(addressBookFileId));
        }
        else if (addressBookFileId instanceof sdk_1.FileId) {
            result.setFileId(addressBookFileId);
        }
        return result;
    }
    /**
     * Creates a new {@link AddressBook} instance. Does not create the file on Hedera File Service!.
     *
     * @param appnetName       Name of the appnet.
     * @param didTopicId       TopicID of the DID topic.
     * @param vcTopicId        Topic ID of the Verifiable Credentials topic.
     * @param appnetDidServers List of appnet API servers.
     * @return The {@link AddressBook}.
     */
    static create(appnetName, didTopicId, vcTopicId, appnetDidServers) {
        const result = new AddressBook();
        result.appnetDidServers = appnetDidServers;
        result.didTopicId = didTopicId;
        result.vcTopicId = vcTopicId;
        result.appnetName = appnetName;
        return result;
    }
    /**
     * Converts this address book file into JSON string.
     *
     * @return The JSON representation of this address book.
     */
    toJSON() {
        return JSON.stringify({
            appnetName: this.appnetName,
            didTopicId: this.didTopicId,
            vcTopicId: this.vcTopicId,
            appnetDidServers: this.appnetDidServers
        });
    }
    getAppnetName() {
        return this.appnetName;
    }
    setAppnetName(appnetName) {
        this.appnetName = appnetName;
    }
    getDidTopicId() {
        return this.didTopicId;
    }
    setDidTopicId(didTopicId) {
        this.didTopicId = didTopicId;
    }
    getVcTopicId() {
        return this.vcTopicId;
    }
    setVcTopicId(vcTopicId) {
        this.vcTopicId = vcTopicId;
    }
    getAppnetDidServers() {
        return this.appnetDidServers;
    }
    setAppnetDidServers(appnetDidServers) {
        this.appnetDidServers = appnetDidServers;
    }
    getFileId() {
        return this.fileId;
    }
    setFileId(fileId) {
        this.fileId = fileId;
    }
}
exports.AddressBook = AddressBook;
