// @hashgraph/did-sdk-js/dist/identity/hcs/message.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Message = void 0;
const long_1 = __importDefault(require("long"));
const timestamp_utils_1 = require("../../utils/timestamp-utils");
class Message {
    constructor() {
        this.timestamp = timestamp_utils_1.TimestampUtils.now();
    }
    getTimestamp() {
        return this.timestamp;
    }
    toJsonTree() {
        const result = {};
        result.timestamp = timestamp_utils_1.TimestampUtils.toJSON(this.timestamp);
        return result;
    }
    toJSON() {
        return JSON.stringify(this.toJsonTree());
    }
    static fromJsonTree(tree, result) {
        if (!result)
            result = new Message();
        result.timestamp = timestamp_utils_1.TimestampUtils.fromJson(tree.timestamp);
        return result;
    }
    static fromJson(json) {
        return Message.fromJsonTree(JSON.parse(json));
    }
}
exports.Message = Message;
Message.serialVersionUID = long_1.default.fromInt(1);
