// @hashgraph/did-sdk-js/dist/identity/hcs/serializable-mirror-consensus-response.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SerializableMirrorConsensusResponse = void 0;
const long_1 = __importDefault(require("long"));
const arrays_utils_1 = require("../../utils/arrays-utils");
const timestamp_utils_1 = require("../../utils/timestamp-utils");
class SerializableMirrorConsensusResponse {
    constructor(response) {
        this.consensusTimestamp = response.consensusTimestamp;
        this.message = response.contents;
        this.runningHash = response.runningHash;
        this.sequenceNumber = response.sequenceNumber;
    }
    toString() {
        return "ConsensusMessage{"
            + "consensusTimestamp=" + timestamp_utils_1.TimestampUtils.toJSON(this.consensusTimestamp)
            + ", message=" + arrays_utils_1.ArraysUtils.toString(this.message)
            + ", runningHash=" + arrays_utils_1.ArraysUtils.toString(this.runningHash)
            + ", sequenceNumber=" + this.sequenceNumber.toNumber()
            + '}';
    }
    toJsonTree() {
        const result = {};
        result.consensusTimestamp = {
            seconds: this.consensusTimestamp.seconds,
            nanos: this.consensusTimestamp.nanos
        };
        result.message = this.message.toString();
        result.runningHash = this.runningHash.toString();
        result.sequenceNumber = this.sequenceNumber.toString();
        return result;
    }
    toJSON() {
        return JSON.stringify(this.toJsonTree());
    }
}
exports.SerializableMirrorConsensusResponse = SerializableMirrorConsensusResponse;
SerializableMirrorConsensusResponse.serialVersionUID = long_1.default.fromInt(1);
